glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,227 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+//! Module for registering shared types for Rust types.
+
+use crate::translate::*;
+
+pub trait RefCounted: Clone + Sized + 'static {
+    /// The inner type
+    type InnerType;
+
+    /// The function used to increment the inner type refcount
+    fn ref_(this: *const Self::InnerType) -> *const Self::InnerType;
+
+    /// Converts the RefCounted object to a raw pointer to InnerType
+    fn into_ptr(self) -> *const Self::InnerType;
+
+    /// Converts a raw pointer to InnerType to a RefCounted object
+    fn from_ptr(this: *const Self::InnerType) -> Self;
+}
+
+impl<T> RefCounted for std::sync::Arc<T>
+where
+    T: 'static,
+{
+    type InnerType = T;
+
+    fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+        unsafe {
+             std::sync::Arc::increment_strong_count(this);
+         }
+         this
+    }
+
+    fn into_ptr(self) -> *const Self::InnerType {
+        std::sync::Arc::into_raw(self)
+    }
+
+    fn from_ptr(this: *const Self::InnerType) -> Self {
+        unsafe { std::sync::Arc::from_raw(this) }
+    }
+}
+
+impl<T> RefCounted for std::rc::Rc<T>
+where
+    T: 'static,
+{
+    type InnerType = T;
+
+    fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+        unsafe {
+            use std::mem::ManuallyDrop;
+            let this_rc = ManuallyDrop::new(std::rc::Rc::from_raw(this));
+            std::rc::Rc::into_raw(ManuallyDrop::take(&mut this_rc.clone()))
```

Comment:
You can use `increment_strong_count` here too

Reply:
For some reason, it seems `increment_strong_count` [is not implemented for `Rc`](https://doc.rust-lang.org/std/rc/struct.Rc.html) in rust 1.51.0

Reply:
Indeed. 1.53: https://github.com/rust-lang/rust/pull/83476